test/e2e/federated-secret.go

```diff
@@ -18,69 +18,161 @@ package e2e
 			}
+			clusters = map[string]*cluster{}
+			_ = registerClusters(clusters, UserAgentName, federationName, f)
+		})
+
+		AfterEach(func() {
+			unregisterClusters(clusters, f)
 		})
 
 		It("should be created and deleted successfully", func() {
 			framework.SkipUnlessFederated(f.Client)
-
 			nsName := f.FederationNamespace.Name
-			secret := createSecretOrFail(f.FederationClientset_1_4, nsName)
-			By(fmt.Sprintf("Creation of secret %q in namespace %q succeeded.  Deleting secret.", secret.Name, nsName))
-			// Cleanup
-			err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
-			framework.ExpectNoError(err, "Error deleting secret %q in namespace %q", secret.Name, secret.Namespace)
-			By(fmt.Sprintf("Deletion of secret %q in namespace %q succeeded.", secret.Name, nsName))
-		})
+			secret, _ := createSecretOrFail(f.FederationClientset_1_4, nsName)
 
+			defer func() { // Cleanup
+				By(fmt.Sprintf("Deleting secret %q in namespace %q", secret.Name, nsName))
+				err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
+				framework.ExpectNoError(err, "Error deleting secret %q in namespace %q", secret.Name, nsName)
+			}()
+			// wait for secret shards being created
+			waitForSecretShardsOrFail(nsName, secret, clusters)
+			secret, _ = updateSecretOrFail(f.FederationClientset_1_4, nsName)
+			waitForSecretShardsUpdatedOrFail(nsName, secret, clusters)
+		})
 	})
 })
 
-func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) *v1.Secret {
-	if clientset == nil || len(namespace) == 0 {
-		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteSecretOrFail: clientset: %v, namespace: %v", clientset, namespace))
-	}
-	By(fmt.Sprintf("Creating federated secret %q in namespace %q", FederatedSecretName, namespace))
-
+func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) (*v1.Secret, error) {
 	secret := &v1.Secret{
 		ObjectMeta: v1.ObjectMeta{
 			Name: FederatedSecretName,
 		},
 	}
-
-	By(fmt.Sprintf("Trying to create secret %q in namespace %q", secret.Name, namespace))
+	By(fmt.Sprintf("Creating secret %q in namespace %q", secret.Name, namespace))
 	_, err := clientset.Core().Secrets(namespace).Create(secret)
-	framework.ExpectNoError(err, "Creating secret %q in namespace %q", secret.Name, namespace)
+	framework.ExpectNoError(err, "Failed to create secret %s", secret.Name)
 	By(fmt.Sprintf("Successfully created federated secret %q in namespace %q", FederatedSecretName, namespace))
-	return secret
+	return secret, err
+}
+
+func updateSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) (newSecret *v1.Secret, err error) {
+	secret := &v1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name: UpdatedFederatedSecretName,
+		},
+	}
+
+	for MaxRetriesOnFederatedApiserver := 0; MaxRetriesOnFederatedApiserver < 3; MaxRetriesOnFederatedApiserver++ {
```

Comment:
MaxRetries could be a constant at the top alongside FederatedSecretTimeout and use a variable retryCount here?